library/common/http/dispatcher.h

```diff
@@ -32,10 +32,48 @@ class Dispatcher : public Logger::Loggable<Logger::Id::http> {
    * @return envoy_stream_t handle to the stream being created.
    */
   envoy_status_t startStream(envoy_stream_t stream, envoy_observer observer);
+
+  /**
+   * Send headers over an open HTTP stream. This method can be invoked once and needs to be called
+   * before send_data.
+   * @param stream, the stream to send headers over.
+   * @param headers, the headers to send.
+   * @param end_stream, supplies whether this is headers only.
+   * @return envoy_status_t, the resulting status of the operation.
+   */
   envoy_status_t sendHeaders(envoy_stream_t stream, envoy_headers headers, bool end_stream);
+
+  /**
+   * Send data over an open HTTP stream. This method can be invoked multiple times.
+   * @param stream, the stream to send data over.
+   * @param data, the data to send.
+   * @param end_stream, supplies whether this is the last data in the stream.
+   * @return envoy_status_t, the resulting status of the operation.
+   */
   envoy_status_t sendData(envoy_stream_t stream, envoy_data data, bool end_stream);
-  envoy_status_t sendMetadata(envoy_stream_t stream, envoy_headers headers, bool end_stream);
-  envoy_status_t sendTrailers(envoy_stream_t stream, envoy_headers headers);
+
+  /**
+   * Send metadata over an HTTP stream. This method can be invoked multiple times.
+   * @param stream, the stream to send metadata over.
+   * @param metadata, the metadata to send.
+   * @return envoy_status_t, the resulting status of the operation.
+   */
+  envoy_status_t sendMetadata(envoy_stream_t stream, envoy_headers headers);
+
+  /**
+   * Send trailers over an open HTTP stream. This method can only be invoked once per stream.
+   * Note that this method implicitly ends the stream.
+   * @param stream, the stream to send trailers over.
+   * @param trailers, the trailers to send.
+   * @return envoy_status_t, the resulting status of the operation.
+   */
+  envoy_status_t sendTrailers(envoy_stream_t stream, envoy_headers trailers);
+
+  /**
+   * Reset an open HTTP stream.
```

Comment:
It'd be nice to elaborate a bit on what this means for the underlying stream handle from the consumer's perspective